### PR TITLE
Iter5

### PR DESCRIPTION
**Pull request by justEngineer**

### CODE REVIEW

File: cmd/agent/config.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	"strconv"
+)
+
+type ClientConfig struct {
+	endpoint       string
+	reportInterval uint64
+	pollInterval   uint64
+}
+
+func GetClientConfig() ClientConfig {
```
Comment (arttet) [inline cmd/agent/config.go:16]:
```suggestion
func Parse() ClientConfig {
```

File: cmd/agent/main.go [hunk 2]
```diff
@@ -1,3 +1,107 @@
...
-func main() {}
+import (
+	"math/rand"
+	"net/http"
+	"runtime"
+	"strconv"
+	"time"
+)
+
+type MemStorage struct {
```
Comment (arttet) [inline cmd/agent/main.go:11]: CHANGE REQUEST: please move it to another package

File: cmd/agent/main.go [hunk 3]
```diff
@@ -1,3 +1,107 @@
...
+	// указаны некоторые поля структуры
+	gauge   map[string]float64
+	counter map[string]int64
+}
+
+var MetricStorage MemStorage
```
Comment (arttet) [inline cmd/agent/main.go:17]: CHANGE REQUEST: please don't use global variables

File: cmd/agent/main.go [hunk 4]
```diff
@@ -1,3 +1,107 @@
...
+		getMetrics()
+		*pollTimeoutSec = 0
+	}
+	if *sendTimeoutSec == int(config.reportInterval) {
+		sendMetrics(config)
+		*sendTimeoutSec = 0
+	}
+}
+
+func main() {
```
Comment (arttet) [inline cmd/agent/main.go:96]: CHANGE REQUEST: only `main` should be here.

File: cmd/server/main.go [hunk 5]
```diff
@@ -1,3 +1,173 @@
...
+		// устанавливаем заголовок Content-Type
+		// для передачи клиенту информации, кодированной в JSON
+		w.Header().Set("Content-Type", "text/plain")
+		// устанавливаем код 200
+		w.WriteHeader(http.StatusOK)
+
+	}
+}
+
+const metricList = `<table>
```
Comment (arttet) [inline cmd/server/main.go:103]: CHANGE REQUEST: it should be a template file

File: cmd/server/main.go [hunk 6]
```diff
@@ -1,3 +1,173 @@
...
+			w.WriteHeader(http.StatusOK)
+			return
+		} else {
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+	}
+}
+
+func main() {
```
Comment (arttet) [inline cmd/server/main.go:157]: CHANGE REQUEST: only `main` should be here.

**Review by arttet — CHANGES_REQUESTED** [review]
Please add the following actions in CI:
- https://about.codecov.io/blog/getting-started-with-code-coverage-for-golang/
- https://github.com/golangci/golangci-lint

File: cmd/agent/config.go (hunk 1 above)
Comment (justEngineer) [reply cmd/agent/config.go]: done

File: cmd/agent/main.go (hunk 2 above)
Comment (justEngineer) [reply cmd/agent/main.go]: done

File: cmd/agent/main.go (hunk 3 above)
Comment (justEngineer) [reply cmd/agent/main.go]: done

File: cmd/agent/main.go (hunk 4 above)
Comment (justEngineer) [reply cmd/agent/main.go:10]: done

File: cmd/server/main.go (hunk 5 above)
Comment (justEngineer) [reply cmd/server/main.go]: done

File: cmd/server/main.go (hunk 6 above)
Comment (justEngineer) [reply cmd/server/main.go:15]: done

File: cmd/agent/config.go (hunk 1 above)
Comment (justEngineer) [reply cmd/agent/config.go]: done

File: internal/http/server/server_handler.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+	"slices"
+	"strconv"
+	"strings"
+
+	"github.com/go-chi/chi/v5"
+
+	storage "github.com/justEngineer/go-metrics-service/internal"
+)
+
+func GetMetric(storage *storage.MemStorage) http.HandlerFunc {
```
Comment (arttet) [inline internal/http/server/server_handler.go:17]:
```suggestion
type Handler struct {
    storage *storage.MemStorage
}

func (h *Handler)GetMetric(w http.ResponseWriter, r *http.Request) {
```

File: internal/http/server/server_handler.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+				return
+			}
+			wd, err := os.Getwd()
+			if err != nil {
+				panic(err)
+			}
+			tmplFile := filepath.Join(filepath.Dir(wd), "../internal/http/server/main_page_html.tmpl")
+			tmpl, err := template.New(tmplFile).ParseFiles(tmplFile)
+			if err != nil {
+				panic(err)
```
Comment (arttet) [inline internal/http/server/server_handler.go:125]: CHANGE REQUEST: you shouldn't panic here

File: internal/http/server/server_handler.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+			}
+			if len(urlPart[1]) != 0 {
+				http.Error(w, "Wrong URL", http.StatusBadRequest)
+				return
+			}
+			wd, err := os.Getwd()
+			if err != nil {
+				panic(err)
+			}
+			tmplFile := filepath.Join(filepath.Dir(wd), "../internal/http/server/main_page_html.tmpl")
```
Comment (arttet) [inline internal/http/server/server_handler.go:122]: CHANGE REQUEST: this is a magic string

**Review by arttet — CHANGES_REQUESTED** [review]

File: internal/http/server/server_handler.go (hunk 9 above)
Comment (justEngineer) [reply internal/http/server/server_handler.go]: I've added it to the server's config and also add addititional command ine argument. Is that what you want?

File: internal/http/server/server_handler.go (hunk 8 above)
Comment (justEngineer) [reply internal/http/server/server_handler.go]: replaced with sending Internal Error status

File: internal/http/server/server_handler.go (hunk 7 above)
Comment (justEngineer) [reply internal/http/server/server_handler.go]: done

File: cmd/agent/main.go [hunk 10]
```diff
@@ -1,3 +1,21 @@
...
+import (
+	"time"
+
+	storage "github.com/justEngineer/go-metrics-service/internal"
+	client "github.com/justEngineer/go-metrics-service/internal/http/client"
+)
+
+func main() {
+	var MetricStorage storage.MemStorage
+	MetricStorage.Gauge = make(map[string]float64)
```
Comment (arttet) [inline cmd/agent/main.go:12]: CHANGE REQUEST: it should be constructed in `MemStorage`

**Review by arttet — APPROVED** [review]

File: cmd/agent/main.go (hunk 10 above)
Comment (justEngineer) [reply cmd/agent/main.go]: done

**Pull request merged**